### PR TITLE
[search] Fix flickering search results

### PR DESCRIPTION
**Pull request by ygorshenin**
Because of precision errors when pivot rectange is calculated from a position, a significant error was in computations of distance from search results to the pivot rectangle center. This PR resolves this issue, and, as a side effect, resolves a problem with buggy integral to string conversion.

@vng @mpimenov PTAL

### CODE REVIEW

**Comment by ygorshenin** [comment]
MAPSME-293

File: search/search_query.cpp [hunk 1]
```diff
@@ -668,9 +706,17 @@ void Query::MakePreResult2(v2::Geocoder::Params const & params, vector<T> & cont
     for (; e != m_results.end() && e->GetPriority() == lastPriority; ++e)
       ;
 
-    // TODO (@y, @m, @vng): this method is deprecated, need to rewrite
-    // it.
-    random_shuffle(b, e);
+    // The main reason of shuffling here is to select a random subset
+    // from the low-priority results. We're using a linear congruental
```
Comment (mpimenov) [inline search/search_query.cpp]: s/congruental/congruential/

File: search/search_query.cpp [hunk 2]
```diff
@@ -549,9 +587,10 @@ class PreResult2Maker
                        search::v2::RankingInfo & info)
   {
     auto const & preInfo = res.GetInfo();
-    m2::PointD pivot = m_params.m_pivot.Center();
 
-    info.m_distanceToPivot = feature::GetMinDistanceMeters(ft, pivot);
+    auto const & pivot = m_params.m_pivotCenter;
+
+    info.m_distanceToPivot = MercatorBounds::DistanceOnEarth(feature::GetCenter(ft), pivot);
```
Comment (vng) [inline search/search_query.cpp]: Раз уж мы тут считаем feature::GetCenter, то давай эту же точку использовать и в PreResult2 как центр результата поиска (можешь передать ее в конструктор).

File: search/search_query.cpp (hunk 1 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/v2/geocoder.hpp [hunk 3]
```diff
@@ -76,7 +76,13 @@ class Geocoder : public my::Cancellable
     Params();
 
     Mode m_mode;
+
+    // We need to pass both pivot and pivot center because pivot is
+    // usually a rectangle created by radius and center, and due to
+    // precision loss, |m_pivot|.Center() may differ from
+    // |m_pivotCenter|.
     m2::RectD m_pivot;
+    m2::PointD m_pivotCenter;
```
Comment (mpimenov) [inline search/v2/geocoder.hpp]: s/pivotCenter/accuratePivotCenter как вариант

File: search/search_query.cpp (hunk 2 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/v2/geocoder.hpp (hunk 3 above)
Comment (ygorshenin) [inline search/v2/geocoder.hpp]: Done

**Comment by ygorshenin** [comment]
PPPTAL @vng @mpimenov

**Comment by mpimenov** [comment]
LGTM

**Pull request merged**